### PR TITLE
Add yield signals to page projection

### PR DESCRIPTION
**Pull request by highker**
Trunk (with `PageProjection` generated and without checking yield signal):
```
Benchmark                          Mode  Cnt     Score     Error  Units
BenchmarkPageProcessor.compiled   thrpt   50  5266.166 ± 170.086  ops/s
BenchmarkPageProcessor.handCoded  thrpt   50  6046.107 ±  25.990  ops/s
```

With both `PageProjectionOutput` and `PageProjection` generated:
```
Benchmark                          Mode  Cnt     Score    Error  Units
BenchmarkPageProcessor.compiled   thrpt   50  5399.072 ± 71.479  ops/s
BenchmarkPageProcessor.handCoded  thrpt   50  6041.955 ± 32.595  ops/s
```

With `PageProjectionOutput` generated and `PageProjection` hand-coded:
```
Benchmark                          Mode  Cnt     Score    Error  Units
BenchmarkPageProcessor.compiled   thrpt   50  5362.882 ± 45.355  ops/s
BenchmarkPageProcessor.handCoded  thrpt   50  5967.048 ± 81.459  ops/s
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java [hunk 1]
```diff
@@ -296,41 +319,139 @@ private static MethodDefinition generatePageProjectMethod(ClassDefinition classD
...
-        Variable index = scope.declareVariable(int.class, "index");
-
         // reset block builder before using since a previous run may have thrown leaving data in the block builder
         body.append(thisVariable.setField(
                 blockBuilder,
                 thisVariable.getField(blockBuilder).invoke("newBlockBuilderLike", BlockBuilder.class, newInstance(BlockBuilderStatus.class))));
 
+        Variable pageProjectionOutput = scope.declareVariable(PageProjectionOutput.class, "pageProjectionOutput");
+        body.append(pageProjectionOutput.set(newInstance(pageProjectionOutputClass, thisVariable.getField(blockBuilder), session, yieldSignal, page, selectedPositions)))
+                .append(pageProjectionOutput.ret());
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java]: These 3 lines can be simplified to:

`body.append(newInstance(pageProjectionOutputClass, thisVariable.getField(blockBuilder), session, yieldSignal, page, selectedPositions).ret)`

File: presto-main/src/test/java/com/facebook/presto/operator/project/TestDictionaryAwarePageProjection.java [hunk 2]
```diff
@@ -154,43 +188,77 @@ private static DictionaryBlock createDictionaryBlockWithUnusedEntries(int dictio
         return new DictionaryBlock(dictionary, ids);
     }
 
-    private static void testProject(Block block, Class<? extends Block> expectedResultType)
+    private static Block projectWithYield(PageProjectionOutput output, DriverYieldSignal yieldSignal)
+    {
+        for (int i = 0; i < 10; i++) {
+            yieldSignal.setWithDelay(1, executor);
+            yieldSignal.forceYieldForTesting();
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/project/TestDictionaryAwarePageProjection.java]: This doesn't look like useful test. Yielding at the beginning essentially means that each `compute` call will do nothing and return almost immediately. I assume that the reason you wrote this test is to make sure that your code will produce correct results regardless of where it got forced a yield.

I feel like a YieldSignal implementation that tosses a coin after every isSet call should be helpful here.

File: presto-main/src/test/java/com/facebook/presto/operator/project/TestDictionaryAwarePageProjection.java [hunk 3]
```diff
@@ -49,85 +65,103 @@ public void testDelegateMethods()
...
+        testProjectList(effectiveBlock, LongArrayBlock.class, projection, forceYield);
 
-        // last dictionary not effective, so dictionary processing is enabled again
-        testProjectRange(ineffectiveBlock, DictionaryBlock.class, projection);
-        testProjectList(ineffectiveBlock, DictionaryBlock.class, projection);
+        // last dictionary effective, so dictionary processing is enabled again
+        testProjectRange(ineffectiveBlock, DictionaryBlock.class, projection, forceYield);
+        testProjectFastReturnIgnoreYield(ineffectiveBlock, projection);
+        // will bypass yield due to enabled dictionary processing
+        testProjectList(ineffectiveBlock, DictionaryBlock.class, projection, false);
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/project/TestDictionaryAwarePageProjection.java]: I don't understand why you want to ignore yield in certain cases.

File: presto-main/src/test/java/com/facebook/presto/operator/project/TestPageProcessor.java [hunk 4]
```diff
@@ -325,21 +325,22 @@ public void testRetainedSize()
...
 
         Slice[] slices = new Slice[rows];
         Arrays.fill(slices, Slices.allocate(rows));
         Page inputPage = new Page(new SliceArrayBlock(slices.length, slices));
 
         PageProcessorOutput output = pageProcessor.process(SESSION, yieldSignal, inputPage);
-        for (int i = 0; i < columns - 1; i++) {
+
+        // one yield within a projection and one yield between two projections
+        for (int i = 0; i < columns * 2 - 1; i++) {
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/project/TestPageProcessor.java]: I don't understand this

**Comment by highker** [comment]
Comments addressed; thanks for the review

File: presto-main/src/test/java/com/facebook/presto/operator/project/TestPageProcessor.java [hunk 5]
```diff
@@ -332,25 +332,31 @@ public void testYieldProjection()
...
         Arrays.fill(slices, Slices.allocate(rows));
         Page inputPage = new Page(new SliceArrayBlock(slices.length, slices));
 
         PageProcessorOutput output = pageProcessor.process(SESSION, yieldSignal, inputPage);
+
+        // Test yield signal works for page processor
+        // The purpose of this test is NOT to test the yield signal in page projection; we have other test to cover that
+        // In page processor, we check yield signal after a column has been completed processed
+        // So we want to set yield signal when the column has just finished processing in order to let page processor capture the yield signal when the block is returned
+        // Also, we would like to reset the yield signal before starting to process the next column in order NOT to yield per position inside the column
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/project/TestPageProcessor.java]: Add period (.) to the end of each line.

completed -> completely
remove "want to" "would like to"

**Pull request merged**